### PR TITLE
Fix the router GEMM inaccuracy when using _front_w in Kimi-K3

### PR DESCRIPTION
**Pull request by b8zhong**
## Problem

The Kimi-K3 MoE layer computes the router logits in bf16. The router then
selects the experts from these bf16 logits. The selection is not accurate.

Other implementations use fp32 for this GEMM:

- vLLM sets `out_dtype=torch.float32` on its gate
- the SGLang EP front (`moe_front.fused_front`) uses `out_dtype=torch.float32`

Measurement with the layer-1 gate weight and the fp32 correction bias:

- 5.27% of the tokens get a different top-16 expert set
- 0.33% of the routed weight goes to different experts (mean); the p99 is 6.4%

Previously, as https://github.com/sgl-project/sglang/pull/29783 shows in GLM-5.2, this is not safe, and can drop the SWE-Bench Verified score by 79 -> 67. In fact since this path is gated by num_tokens, it will be really difficult to diagnose bugs in the future (when running evals quickly with large batch sizes, vs real world use at small batch sizes)

## The `_front_w` fusion

Three projections in the MoE layer read the same `hidden_states`:

| projection | weight shape at TP8 |
|---|---|
| `shared_experts.gate_up_proj` | `[1536, 7168]` |
| `gate` (the router) | `[896, 7168]` |
| `routed_expert_down_proj` | `[3584, 7168]` |

`_merge_front_weights()` concatenates the three weights along the row axis into
one tensor, `_front_w [6016, 7168]`. One GEMM then replaces three GEMMs. Each
module keeps a view of the merged tensor, so the memory does not increase.

```
hidden_states [M, 7168] --> _front_w [6016, 7168] --> fused [M, 6016]
                                                        |
      cols    0:1536  gate_up  --> SiTU --> shared down_proj
      cols 1536:2432  logits   --> router (sigmoid + bias + top-16)
      cols 2432:6016  latent   --> mxfp8 quant --> trtllm-gen MoE
```

The fusion saves 11-16 us for each MoE layer. The fusion also causes the problem. `fused` is one tensor, and one tensor has one dtype. Two of the three consumers need bf16 (SiTU, MXFP8 quant). Thus the router got bf16, which works functionally but not numerically correct.

## The fix

The GEMM writes fp32. The two other consumers read fp32 directly. No cast pass
and no new kernel are necessary.

- **`cutedsl_bf16_gemm`** can now write an fp32 destination. The output dtype
  can be bf16 or fp32. The accumulator was already fp32, and the store is a
  plain `st.global`. Only the epilogue cast goes away. The operands stay bf16.
  This keeps the TGV kernel at m <= 8, where it is 3 us faster than cuBLAS.
- **`situ_and_mul`** takes different input and output types. It converted to
  fp32 internally before this change, so a wider input only changes the load
  width.
- **`QuantTrait`** gets an fp32 input path. `kVecSize` is now 16 elements and no longer depends on
  the input width, because `route_quant_fused` derives its CTA geometry from
  `kNumLanes`.
- **`route_quant_fused`** is templated on the activation dtype.

The other MoE runners (marlin, triton) take `routed_input` in bf16. These
runners already made a dense copy of that strided slice. The copy now also does
the cast. Only a one-token batch adds a launch, because its slice was already
dense. Thus every CUDA runner gets the fp32 router, on SM90 and on Blackwell.

`routed_input` stays bit-identical to the previous output, because
`bf16(tgv_fp32) == tgv_bf16`. The GEMM output for `gate_up` is also
bit-identical, but SiTU now reads the fp32 values. Thus the shared-expert
activation changes. It becomes more accurate: the maximum relative error
against an fp32 reference decreases from 3.9e-03 to 2.5e-03.

## Results

Time for each MoE layer at TP8 decode shapes. This is the complete front: the
GEMM, SiTU, and the route/quant. CUPTI, with CUDA graph capture.

| tokens | 1 | 8 | 32 | 64 | 128 |
|---|---|---|---|---|---|
| bf16 (before) | 25.28 | 25.41 | 28.42 | 27.27 | 34.82 |
| fp32 (after) | 24.35 | 24.74 | 27.46 | 26.40 | 33.57 |
| delta | -0.93 | -0.67 | -0.96 | -0.87 | -1.28 |

The accurate path is faster than the path it replaces. The gain comes from
SiTU: an fp32 input halves the vector for each lane, which doubles the number
of threads on a latency-bound kernel.

The other runners pay for the cast of `routed_input`. Time for the front GEMM
and that cast, at the same shapes:

| tokens | 1 | 2 | 4 | 8 | 16 | 32 | 64 | 128 |
|---|---|---|---|---|---|---|---|---|
| bf16 + copy | 20.45 | 23.58 | 23.58 | 24.06 | 25.47 | 25.95 | 24.74 | 25.66 |
| fp32 + cast | 21.95 | 23.17 | 23.58 | 23.42 | 25.60 | 26.11 | 24.86 | 26.34 |

Only one token shows a cost, because the cast is a new launch there.

## Notes

- ROCm keeps the bf16 front. `torch.mm` with `out_dtype` and the aiter routing
  path are not verified for the fp32 chain.
- `test_per_token_group_quant_8bit_v2.py::test_v2_jit_matches_aot` fails 66
  cases on main. This failure is not related to this PR. I confirmed it with
  the changes removed.

---
### CI States

Latest PR Test (Base): :white_check_mark: [Run #31261862166](https://github.com/sgl-project/sglang/actions/runs/31261862166)
Latest PR Test (Extra): :x: [Run #31261862100](https://github.com/sgl-project/sglang/actions/runs/31261862100)

### CODE REVIEW

**Review by BBuf — APPROVED** [review]
Any end2end acc can be reported?

**Comment by b8zhong** [comment]
@Fridge003 I'm testing this pr with perf and acc numbers. Also the approach might be reconsidered, to avoid adding complexity like this, instead we might unfuse router, but it requires replacing with the tinyN GEMMs for perf, since currently multi-streaming or serializing it both causes bad E2E (-4-5%). The current PR numbers is around ~0.5% at BS = 8 (should be noise), but we need to verify across token counts

**Comment by b8zhong** [comment]
Here is 1K/1K perf data on TP-8 B300. Seems within noise

| bs | main | FP32 | Δ |
|---:|---:|---:|---:|
| 1 | 7.92 | 7.94 | +0.26% |
| 2 | 8.70 | 8.71 | +0.03% |
| 4 | 9.83 | 9.84 | +0.03% |
| 8 | 12.44 | 12.35 | -0.68% |
| 12 | 15.04 | 14.96 | -0.54% |
| 16 | 16.74 | 16.64 | -0.62% |
| 24 | 19.52 | 19.58 | +0.31% |
| 32 | 21.77 | 21.82 | +0.20% |
| 40 | 24.22 | 24.20 | -0.10% |
| 48 | 25.75 | 25.73 | -0.07% |
| 52 | 27.27 | 27.25 | -0.06% |
| mean | | | -0.11% |
